web/src/message_store.ts

```diff
@@ -197,6 +197,19 @@ export function get_pm_full_names(user_ids: number[]): string {
     return names.join(", ");
 }
 
+export function get_last_visible_msg_id_in_topic(
+    stream_name: string,
+    topic: string,
+): number | undefined {
+    const channel_topic_msgs: number[] = [];
+    for (const [msg_id, msg] of stored_messages.entries()) {
+        if (msg.type === "stream" && msg.display_recipient === stream_name && msg.topic === topic) {
+            channel_topic_msgs.push(msg_id);
+        }
+    }
+    return Math.max(...channel_topic_msgs);
+}
```

Comment:
This algorithm doesn't seem reliable; `message_store` has no guarantee that it contains the very latest messages.

The `recent_view_data` module should be our most correct resource for this; `stream_topic_history` might be more accurate, but it has some live-update issues documented in a big TODO comment in that file. @amanagr FYI just because we might be able to tweak this PR if your efforts on message-moving end up removing that TODO.

Reply:
thanks for letting me know! I'll update it.

Reply:
seems like `recent_view_data` sometimes does not contain the topics available in left sidebar -- it gives out undefined when I try to access their `last_msg_id`

Reply:
Yeah, `stream_topic_history` is what you need if someone has clicked `show all topics`, so maybe we need to work from that instead.

Reply:
I agree, `stream_topic_history` should be our source of data here.

Reply:
updated to use `stream_topic_history`.